### ISSUE TITLE
PMAPs interface transition

### ISSUE THREAD

**Issue by jacg** [next-exp/IC#407]
This issue exists as a focal point for the activity surrounding the PMAPs interface transition.

The work is to be done in a sequence of several PRs and issues:

1. [x] **INTERFACE** A module with just the interface and no implementation:  #394 **COMPLETED**
2. [x] **IMPLEMENTATION** Tests and corresponding *pure Python* implementation behind the interface approved in step 1: #406 **IN PROGRESS**
3. [x] **THE BIG REFACTOR** Adapt all existing IC code to use the new interface. Remove all trace of the old interface and implementation from `HEAD`: **AWAITING COMPLETION OF #406**
4. [ ] **DOCUMENT** As this is the first official, stable part of our API, it needs to de *documented* properly. Unlike anything else we've written so far, we are committed to keeping this stable! **CAN BE STARTED RIGHT NOW!** Don't forget to use a branch whose name starts with `docs.`!
5. [ ] **PROFILE** Publish results. Analyse. If *and only if* the profile results suggest that optimizing the implementation might result in overall improvement of performance, do we even think of doing the next step. **AWATING COMPLETION OF STEP 3**
6. [ ] **OPTIMIZE** If Cython is used, no Cython modules should appear in the interface: Cythonized functions should be imported into the pure Python interface modules, from which clients will import them, not knowing or caring whether they are implemented in Cython or not. **AWAITING PERMISSION FROM STEP 5.**

The last 2 steps can be delayed indefinitely.  The documentation step should be completed before the January sprint!

**Comment by jacg** [comment]
Is The Big Refactor taking place right now? or is it next week? When do we expect it to be finished? Please remember that we have to budget time for preparing the documentation **before** the January sprint!

**Comment by abotas** [comment]
The big refactor is underway and hopefully almost done!

**Comment by abotas** [comment]
It will definitely be finished this week I think. Hopefully in the next couple days.

@gonzaponte and I are putting this in my branch adapt to new pmaps

**Comment by jacg** [comment]
#406 was supposed to have been approved **before** starting The Big Refactor.

If you guys are doing The Big Refactor right now, then I'll incorporate my comments into #406 and approve and merge it, unless I hear something to the contrary.

I infer that in the meeting you had last week everyone was extatically happy with #406 and gave you the go-ahead.

**Comment by gonzaponte** [comment]
> #406 was supposed to have been approved before starting The Big Refactor.

I know, but we started to work on it before the meeting, which was supposed to be the review. 

Furthermore, I find useful to have some overlap between both. At the end of the day, it doesn't matter whether the PR has been approved or not as long as we have the same interface and the product remains stable. For instance, we've found during TBR that we hadn't made the safest implementation (nor the best tests). We were able to initialize the classes with wrong inputs and nothing complained. We can always go back and submit a patch in another PR, but it is nice to have the possibility of including it in the "original" implementation. What do you think?

**Comment by jacg** [comment]
> it is nice to have the possibility of including it in the "original" implementation.

Fine. Carry on. Probably a good idea, actually.

I was hoping to be able to rely on it being stable while you guys carry on with TBR, but this is probably more useful.

**Comment by gonzaponte** [comment]
Wow, I was expecting a big "bronca"!
